src/custom_node_library_manager.cpp

```diff
@@ -0,0 +1,79 @@
+//*****************************************************************************
+// Copyright 2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "custom_node_library_manager.hpp"
+
+#include <utility>
+
+#include <dlfcn.h>
+
+#include "logging.hpp"
+
+namespace ovms {
+
+Status CustomNodeLibraryManager::loadLibrary(const std::string& name, const std::string& basePath) {
+    SPDLOG_LOGGER_INFO(modelmanager_logger, "Loading custom node library name: {}; base_path: {}", name, basePath);
+
+    if (libraries.count(name) == 1) {
+        SPDLOG_LOGGER_INFO(modelmanager_logger, "Custom node library name: {} is already loaded", name);
+        return StatusCode::NODE_LIBRARY_ALREADY_LOADED;
+    }
+
+    void* handle = dlopen(basePath.c_str(), RTLD_LAZY | RTLD_LOCAL);
+    char* error = dlerror();
+    if (handle == NULL) {
+        SPDLOG_LOGGER_INFO(modelmanager_logger, "Library name: {} failed to open base_path: {} with error: {}", name, basePath, error);
+        return StatusCode::NODE_LIBRARY_LOAD_FAILED;
+    }
+
+    execute_fn execute = (execute_fn)dlsym(handle, "execute");
+    error = dlerror();
+    if (error || execute == nullptr) {
+        SPDLOG_LOGGER_INFO(modelmanager_logger, "Library name: {} with error: {}", name, error);
+        dlclose(handle);
+        return StatusCode::NODE_LIBRARY_LOAD_FAILED;
+    }
+
+    release_fn releaseBuffer = (release_fn)dlsym(handle, "releaseBuffer");
+    error = dlerror();
+    if (error || releaseBuffer == nullptr) {
+        SPDLOG_LOGGER_INFO(modelmanager_logger, "Library name: {} with error: {}", name, error);
+        dlclose(handle);
+        return StatusCode::NODE_LIBRARY_LOAD_FAILED;
+    }
+
+    release_fn releaseTensors = (release_fn)dlsym(handle, "releaseTensors");
+    error = dlerror();
+    if (error || releaseBuffer == nullptr) {
```

Comment:
Shouldn;t we return different kind of error for missing different part of library?

Reply:
@atobiszei I don't think so. We will never use this information in application. We will just skip library if any error appears. Therefore I think separate log for each part of library missing is enough.

Reply:
That's fair for retCode but log message to report what stage failed in logs would be usefull still. User can know that the library is missing execute/releaseBuffer or something else?